examples/miscellaneous/plot_outlier_detection_bench.py

```diff
@@ -3,193 +3,414 @@
 ==========================================
 
-This example benchmarks outlier detection algorithms, :ref:`local_outlier_factor`
-(LOF) and :ref:`isolation_forest` (IForest), using ROC curves on
-classical anomaly detection datasets. The algorithm performance
-is assessed in an outlier detection context:
+This example compares two outlier detection algorithms, namely
+:ref:`local_outlier_factor` (LOF) and :ref:`isolation_forest` (IForest), on
+real-world datasets available in :class:`sklearn.datasets`. The goal is to show
+that different algorithms perform well on different datasets.
 
-1. The algorithms are trained on the whole dataset which is assumed to
-contain outliers.
+The algorithms are trained in an outlier detection context:
 
-2. The ROC curve from :class:`~sklearn.metrics.RocCurveDisplay` is computed
-on the same dataset using the knowledge of the labels.
+1. The ROC curves are computed using knowledge of the ground-truth labels
+and displayed using :class:`~sklearn.metrics.RocCurveDisplay`.
 
+2. The performance is assessed in terms of the ROC-AUC.
 """
 
 # Author: Pharuj Rajborirug <pharuj.ra@kmitl.ac.th>
+#         Arturo Amor <david-arturo.amor-quiroz@inria.fr>
 # License: BSD 3 clause
 
-print(__doc__)
+# %%
+# Dataset preprocessing and model training
+# ========================================
+#
+# Different outlier detection models require different preprocessing. In the
+# presence of categorical variables,
+# :class:`~sklearn.preprocessing.OrdinalEncoder` is often a good strategy for
+# tree-based models such as :class:`~sklearn.ensemble.IsolationForest`, whereas
+# neighbors-based models such as :class:`~sklearn.neighbors.LocalOutlierFactor`
+# would be impacted by the ordering induced by ordinal encoding. To avoid
+# inducing an ordering, on should rather use
+# :class:`~sklearn.preprocessing.OneHotEncoder`.
+#
+# Neighbors-based models may also require scaling of the numerical features (see
+# for instance :ref:`neighbors_scaling`). In the presence of outliers, a good
+# option is to use a :class:`~sklearn.preprocessing.RobustScaler`.
+#
+# The following `fit_predict` function returns the average outlier score of X.
+# In this example we set `n_neighbors` to match the expected number of anomalies
+# `expected_n_anomalies = n_samples * expected_anomaly_frac`. This is a good
+# heuristic as long as the proportion of outliers is not very low, as
+# `n_neighbors` should be at least greater than the number of samples in the
+# less populated cluster (see
+# :ref:`sphx_glr_auto_examples_neighbors_plot_lof_outlier_detection.py`).
+
+from sklearn.neighbors import LocalOutlierFactor
+from sklearn.ensemble import IsolationForest
+from sklearn.preprocessing import (
+    OneHotEncoder,
+    OrdinalEncoder,
+    RobustScaler,
+)
+from sklearn.compose import ColumnTransformer
+from sklearn.pipeline import make_pipeline
+from time import perf_counter
+
+
+def fit_predict(X, model_name, expected_anomaly_frac, categorical_columns=()):
```

Comment:
Reading a bit more about the example, I would decouple the creation of the estimator and the real `fit_predict`:

```python
# %%
# ... text about the desired preprocessing and modelling
from sklearn.neighbors import LocalOutlierFactor
from sklearn.ensemble import IsolationForest
from sklearn.preprocessing import (
    OneHotEncoder,
    OrdinalEncoder,
    RobustScaler,
)
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import make_pipeline


def make_estimator(name, categorical_columns=None, **kwargs):
    """Create an outlier detection estimator based on its name."""
    if name == "LOF":
        outlier_detector = LocalOutlierFactor(**kwargs)
        if categorical_columns is None:
            preprocessor = RobustScaler()
        else:
            preprocessor = ColumnTransformer(
                transformers=[("categorical", OneHotEncoder(), categorical_columns)],
                remainder=RobustScaler(),
            )
    else:  # name == "IForest"
        outlier_detector = IsolationForest(**kwargs)
        if categorical_columns is None:
            preprocessor = None
        else:
            ordinal_encoder = OrdinalEncoder(
                handle_unknown="use_encoded_value", unknown_value=-1
            )
            preprocessor = ColumnTransformer(
                transformers=[
                    ("categorical", ordinal_encoder, categorical_columns),
                ],
                remainder="passthrough",
            )

    return make_pipeline(preprocessor, outlier_detector)


# %%
# ... text about the `fit_predict`
# %%
from time import perf_counter


def fit_predict(estimator, X):
    tic = perf_counter()
    if estimator[-1].__class__.__name__ == "LocalOutlierFactor":
        estimator.fit(X)
        y_pred = estimator[-1].negative_outlier_factor_
    else:  # "IsolationForest"
         y_pred = estimator.fit(X).decision_function(X)
    toc = perf_counter()
    print(f"Duration for {model_name}: {toc - tic:.2f} s")
    return y_pred

Reply:
In the code above, it means that we will show how to define `n_neighbors` each time but we can also pass `random_state` when requesting an isolation forest.

Reply:
As `LocalOutlierFactor` and `IsolationForest` don't share `kwargs`, the suggested `make_estimator` function would only work inside `if`-statements that make more boilerplate code. The idea was to have a function to handle the cases to avoid this.

Reply:
I could still pass `kwargs` to the `LocalOutlierFactor` and set the `random_state` directly in the construction of the `IsolationForest`, though I'm aware this is not a good practice. WDYT?